### PR TITLE
Warn user when path is invalid

### LINKED ISSUE #1487: ensure local path is valid when creating repository
Supersedes #1067

### Description

We should provide a helpful message when the user has entered a known invalid path, before trying to create it.

### Version

**GitHub Desktop version:** 0.5.1

**OS version:** `Microsoft Windows [Version 10.0.15063]`

### Steps to Reproduce

1. Launch `File` -> `New Repository`
1. Change the local path to `>`

**Expected behaviour:** some sort of validation that the path is not valid

**Actual behaviour:** no message, can continue with creating the repository

**Reproduces how often:** 100%

#### Additional Information

The error you get differs per-platform. Here's what it is on Windows:

And on macOS:

ecc58010-109b-11e7-9321-dd9945c18ea0

The other downside of this current behaviour is that it occurs _after_ the form has been dismissed - so there's no way for the user to recover - they need to go through the process all over again.

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
Fixes #1487

This changes the behavior of the **Create a New Repository** 
* ~The _Create Repository_ button is now disabled if the path entered is invalid~
* A warning message appears under the _Local Path_ input after user attempts to create repo if the path is invalid

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
@shiftkey do you have any better ideas for what they warning text should be? I think it's a bit wordy ATM.

File: app/src/ui/add-repository/create-repository.tsx [hunk 1]
```diff
@@ -81,11 +83,25 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
...
 
-  private onPathChanged = (event: React.FormEvent<HTMLInputElement>) => {
-    const path = event.currentTarget.value
-    this.setState({ ...this.state, path })
+  private onPathChanged = async (event: React.FormEvent<HTMLInputElement>) => {
+    try {
+      const path = event.currentTarget.value
+
+      this.setState({ ...this.state, path })
+      await this.ensureDirectory(path)
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Just a heads up: `FSE.ensureDir` will probably fire for each keystroke here, and as per [the docs](https://github.com/jprichardson/node-fs-extra/blob/master/docs/ensureDir.md) you might create a lot of unnecessary folders.

> Ensures that the directory exists. **If the directory structure does not exist, it is created. **

File: app/src/ui/add-repository/create-repository.tsx [hunk 2]
```diff
@@ -265,8 +283,25 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
...
+  private renderInvalidPathWarning() {
+    if (this.state.isValidPath) {
+      return null
+    }
+
+    return (
+      <Row className='warning-helper-text'>
+        <Octicon symbol={OcticonSymbol.alert} />
+        <p>
+          Cannot create repository at selected path. <br/>
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Not sure I can make this less wordy, but what about something like this:

> Directory could not be created at this path.
> Your user account may not have permissions to do this.

Or:

> Directory could not be created at this path.
> You may not have permissions to create a directory here.

Comment (iAmWillShepherd) [inline app/src/ui/add-repository/create-repository.tsx]: I like that second one

File: app/src/ui/add-repository/create-repository.tsx (hunk 1 above)
Comment (iAmWillShepherd) [inline app/src/ui/add-repository/create-repository.tsx]: Yes. It does... 😠 

Glad I have time machine.

**Comment by iAmWillShepherd** [comment]
I changed the behavior. This now prevents the error from popping up, but path validity is still determined **after** the user clicks the _Create Repository_ button.

💯

**Comment by niik** [comment]
> I changed the behavior. This now prevents the error from popping up, but path validity is still determined after the user clicks the Create Repository button.

In that case I think this should be presented as a `DialogError` rather than inline to conform with how our other dialogs work.

**Comment by iAmWillShepherd** [comment]
@niik 

> Are you planning on tackling the reverse scenario as well (someone giving the create dialog a path which is already a repository)?

I've started working on that in this PR; however, this PR now depends on #1743 being merged.

File: app/src/ui/add-repository/create-repository.tsx [hunk 3]
```diff
@@ -119,7 +129,12 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
 
     try {
       await this.ensureDirectory(fullPath)
+      this.setState({ ...this.state, isValidPath: true })
     } catch (e) {
+      if (e.code === 'EACCES' && e.errno === -13) {
+        return this.setState({ ...this.state, isValidPath: false })
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Minor thing and unrelated to this change -`Dialog.onSubmit` is the prop we assign `createRepository` to, and it's signature is `() => void`. But `createRepository` has an inferred signature of `() => Promise<void>` due to the places below where we are returning `postError`.

This usage should be fine for the sake of this change, but maybe we should let `Dialog.onSubmit` take a promise-returning function too.

File: app/src/ui/add-repository/create-repository.tsx [hunk 4]
```diff
@@ -265,10 +280,62 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
     )
   }
 
+  private renderInvalidPathWarning() {
+    const isValidPath = this.state.isValidPath
+
+    if (isValidPath === null || isValidPath) {
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Could we extract the `isValidPath === null` check to a more friendly name, so we can understand that we're only really interested in `false` here for rendering the error?

File: app/src/ui/add-repository/create-repository.tsx [hunk 5]
```diff
@@ -265,10 +280,62 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
...
+          Directory could not be created at this path.
+          You may not have permissions to create a directory here.
+      </DialogError>
+    )
+  }
+
+  private renderIsGitRepoWarning() {
+    const isRepo = this.state.isGitRepository
+
+    if (!this.state.path.length || isRepo == null || isRepo) { return null }
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: `===` ?

File: app/src/ui/add-repository/create-repository.tsx [hunk 6]
```diff
@@ -265,10 +280,62 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
...
+
+    return (
+      <Row className='warning-helper-text'>
+        <Octicon symbol={OcticonSymbol.alert} />
+        <p>
+          This directory appears to already be a git repository. Would you like to <LinkButton onClick={this.onAddRepositoryClicked}>add this repository</LinkButton> instead?
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Wording suggestion:

> This directory appears to be a Git repository. Would you like to add this repository instead?

Comment (iAmWillShepherd) [inline app/src/ui/add-repository/create-repository.tsx]: I like that.

**Comment by iAmWillShepherd** [comment]
# Updated Functionality



💯

File: app/src/ui/add-repository/add-existing-repository.tsx [hunk 7]
```diff
@@ -18,6 +18,11 @@ const untildify: (str: string) => string = require('untildify')
 interface IAddExistingRepositoryProps {
   readonly dispatcher: Dispatcher
   readonly onDismissed: () => void
+
+  /** An optional path to prefill the path text box with.
+   * Defaults to the empty string
```
Comment (shiftkey) [inline app/src/ui/add-repository/add-existing-repository.tsx]: Maybe finish off this sentence like this:

> Defaults to the empty string **if not defined**.

File: app/src/ui/add-repository/create-repository.tsx (hunk 6 above)
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Bump: the first sentence can be tweaked:

> This directory appears to be a Git repository.

**Comment by iAmWillShepherd** [comment]
💯

File: app/src/ui/add-repository/create-repository.tsx [hunk 8]
```diff
@@ -272,10 +293,52 @@ export class CreateRepository extends React.Component<ICreateRepositoryProps, IC
...
+  private renderInvalidPathWarning() {
+    const isValidPath = this.state.isValidPath
+    const pathSet = isValidPath !== null
+
+    if (!pathSet || isValidPath) {
+      return null
+    }
+
+    return (
+      <DialogError>
```
Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: Probably out of scope for the PR, but the red background when this is drawn stands out excessively when it's shown inline, rather than as the header. 



versus

screen shot 2017-06-18 at 10 21 19 pm

We can :boot: this to a follow-up PR as we don't have the styles for this currently anyway.

Comment (shiftkey) [inline app/src/ui/add-repository/create-repository.tsx]: #2052
